File: jaeseuk/트리/b11725.py
from sys import stdin, setrecursionlimit
setrecursionlimit(10 ** 9)


def dfs(idx):
    for node in tree[idx]:
        if parents[node] == 0:
            parents[node] = idx
            dfs(node)


n = int(stdin.readline().rstrip())
tree = [[] for _ in range(n + 1)]
parents = [0] * (n + 1)

for _ in range(n - 1):
    a, b = map(int, stdin.readline().rstrip().split())
    tree[a].append(b)
    tree[b].append(a)

dfs(1)

for i in range(2, n + 1):
    print(parents[i])

File: yoongyeong/by_kotlin/src/main/kotlin/tree/B11725.kt
package tree

import java.util.StringTokenizer

// 트리의 부모 찾기

private lateinit var trees: Array<MutableList<Int>>

fun main() {
    val br = System.`in`.bufferedReader()
    val n = br.readLine().toInt()
    trees = Array(n+1) { mutableListOf() }

    repeat(n-1) {
        with(StringTokenizer(br.readLine())) {
            val a = nextToken().toInt(); val b = nextToken().toInt()
            trees[a].add(b); trees[b].add(a)
        }
    }

    val parents = IntArray(n+1)

    dfs(parents, n, 1, 0)

    for (i in 2 .. n) println(parents[i])
}

private fun dfs(parents: IntArray, n: Int, start: Int, parent: Int) {
    parents[start] = parent
    for (node in trees[start]) if (node != parent) dfs(parents, n, node, start)
}